text/0000-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,556 @@
+
+## Drawbacks
+
+- The existing APIs have served Ember well for a long time. Introducing any new change involves some risk, particularly in areas so core to how Ember behaves.
+
+- The proposed API here trades away brevity in favor of clarity, tooling support, and being “normal” JavaScript. `'service:session'` is many fewer characters to type than `{ type: 'service', name: 'session' }`.
+
+## Alternatives
+
+### A new String-based API
+
+Instead of providing an object-based API, we could provide a new string-based API. The `lookup` method, then, would have a signature like this:
+
+```ts
+interface Owner {
+  lookup(type: string, name: string, options: Options);
+}
+```
+
+This has the advantage of being a smaller diff from today’s world, and cutting down on the extra typing of the new API design:
+
+```diff
+- lookup('service:foo')
++ lookup('service', 'foo')
+```
+
+vs.
+
+```diff
+- lookup('service:foo');
++ lookup({ type: 'service', name: 'foo' })
+```
+
+This is a less flexible proposal, and—more importantly—it has a significant downside for usability beyond the simplest case. Supporting namespaces here would require *another* string argument, and functions with multiple arguments of the same type are notoriously easy to misuse. (“Wait, what’s the order? Namespace first or type first? Does namespace come after type, or after name?” etc.)
+
+### String-based API as “sugar”
+
+While this RFC recommends the object-based API as the *primary* API, we could supply the string-based API outlined above for the simple and most common case, so that this—
+
+```js
+lookup('service', 'foo')
+```
+
+—would be equivalent to this:
+
+```js
+lookup({ type: 'service', name: 'foo'})
+```
+
+This could be nice from an ergonomics perspective, and introduces only a small amount of extra complexity in the implementation: the resolver `schemaVersion` key still allows the implementation to distinguish between the `lookup('<type>:<name>')` form and the `lookup(type, name)` form. However, it does require introspection on the arguments passed to the functions.
+
+### Object-based API
+
+Another brief syntax might use the type as the key and the name as its value:
+
+```js
+lookup({ service: 'foo' })
+```
+
+While this is nice from the perspective of the consumer of the API, it muddies the API substantially and makes implementation more complex and with much worse performance: the lookup would have to iterate over all keys on the object passed every time, mapping those keys to the *known* identifer interface and then checking the others against the registry. This also makes the TypeScript implementation considerably more difficult, for similar reasons to the runtime implementation, but at compile time.
```

Comment:
What would it mean to do `lookup({ service: 'foo', route: 'bar' })`!?!?

Reply:
Oh my, I hadn't even thought of that as a downside. 😂 That's *terrible*.